### PR TITLE
[SYCL-MLIR] Define attributes wrapping enums in `SYCLAttributes.td`

### PR DESCRIPTION
**Pull request by victor-eds**
This is a first step towards using attributes in the dialect, e.g., SYCL address spaces.

In order to make use of these, we need to first define proper attributes for each enum we had.

Note this is not yet reachable from `cgeist`.

### CODE REVIEW

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/CMakeLists.txt [hunk 1]
```diff
@@ -7,8 +7,10 @@ add_mlir_interface(SYCLMethodOpInterface)
 add_mlir_doc(SYCLMethodOpInterface SYCLMethodOpInterface Dialects/ -gen-op-interface-docs)
 
 set(LLVM_TARGET_DEFINITIONS SYCLAttributes.td)
-mlir_tablegen(SYCLAttributes.h.inc -gen-enum-decls)
-mlir_tablegen(SYCLAttributes.cpp.inc -gen-enum-defs)
+mlir_tablegen(SYCLEnums.h.inc -gen-enum-decls)
+mlir_tablegen(SYCLEnums.cpp.inc -gen-enum-defs)
+mlir_tablegen(SYCLAttributes.h.inc -gen-attrdef-decls)
+mlir_tablegen(SYCLAttributes.cpp.inc -gen-attrdef-defs)
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/CMakeLists.txt:13]: Adding attributes

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLAttributes.td [hunk 2]
```diff
@@ -13,11 +13,18 @@ include "mlir/IR/EnumAttr.td"
...
     : I32EnumAttr<name, summary, cases> {
+  let genSpecializedAttr = 0;
   let cppNamespace = "::mlir::sycl";
 }
 
+class SYCL_I32EnumAttr<string name, string summary, string mnemonic,
+                       list<I32EnumAttrCase> cases>
+    : EnumAttr<SYCL_Dialect, SYCL_I32Enum<name, summary, cases>, mnemonic> {
+  let assemblyFormat = "`<` $value `>`";
+}
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLAttributes.td:26]: We now generate both enums and attributes for each case. We will be using enums for now, as migration to attributes require many changes.

File: mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp [hunk 3]
```diff
@@ -83,10 +83,12 @@ bool SYCLAddrSpaceCastOp::areCastCompatible(TypeRange inputs,
...
-  unsigned int inputMS = input.getMemorySpaceAsInt();
-  unsigned int outputMS = output.getMemorySpaceAsInt();
-  return ((inputMS == genericAddressSpace) !=
-          (outputMS == genericAddressSpace));
+  auto inputMS = dyn_cast_or_null<AccessAddrSpaceAttr>(input.getMemorySpace());
+  auto outputMS =
+      dyn_cast_or_null<AccessAddrSpaceAttr>(output.getMemorySpace());
+  return (inputMS && outputMS &&
+          (inputMS.getValue() == AccessAddrSpace::GenericAccess) !=
+              (outputMS.getValue() == AccessAddrSpace::GenericAccess));
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp:91]: This is the only place we're using these attributes for now as a proof of concept. We're not generating this operation in `cgeist`, so no change is needed there.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-addrspacecast-to-llvm-typed-pointer.mlir [hunk 4]
```diff
@@ -4,10 +4,13 @@
...
-func.func @PtrCastToGeneric(%arg0: memref<?xi32>) -> memref<?xi32, 4> {
-  %0 = sycl.addrspacecast %arg0 : memref<?xi32> to memref<?xi32, 4>
-  return %0 : memref<?xi32, 4>
+func.func @PtrCastToGeneric(
+    %arg0: memref<?xi32, #sycl.access.address_space<private>>)
+    -> memref<?xi32, #sycl.access.address_space<generic>> {
+  %0 = sycl.addrspacecast %arg0
+      : memref<?xi32, #sycl.access.address_space<private>>
+      to memref<?xi32, #sycl.access.address_space<generic>>
+  return %0 : memref<?xi32, #sycl.access.address_space<generic>>
```
Comment (victor-eds) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-addrspacecast-to-llvm-typed-pointer.mlir:13]: Target-independent code

File: polygeist/lib/Dialect/Polygeist/Transforms/BareMemRefToLLVM.cpp [hunk 5]
```diff
@@ -655,14 +655,23 @@ void mlir::polygeist::populateBareMemRefToLLVMConversionPatterns(
...
 
+        FailureOr<unsigned> addrSpace = converter.getMemRefAddressSpace(type);
+        if (failed(addrSpace)) {
+          emitError(UnknownLoc::get(type.getContext()),
+                    "conversion of memref memory space ")
+              << type.getMemorySpace()
+              << " to integer address space failed. Consider adding memory "
+                 "space conversions.";
+          return Type{};
+        }
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/BareMemRefToLLVM.cpp:666]: Basically took this from the non-bare case [here](https://github.com/intel/llvm/blob/50f1605d2eb1c89ea964d0730521aa532fc7c30e/mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp#L326)

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 6]
```diff
@@ -2325,6 +2326,15 @@ class GroupGetLocalLinearRangePattern
 //===----------------------------------------------------------------------===//
 
 void mlir::populateSYCLToLLVMTypeConversion(LLVMTypeConverter &typeConverter) {
+  typeConverter.addTypeAttributeConversion(
+      [](BaseMemRefType, AccessAddrSpaceAttr addrSpace)
+          -> TypeConverter::AttributeConversionResult {
+        return IntegerAttr::get(
+            IntegerType::get(addrSpace.getContext(), 64),
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2333]: should we use index type here?

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/CMakeLists.txt (hunk 1 above)
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/CMakeLists.txt:13]:
```suggestion
mlir_tablegen(SYCLAttributes.h.inc -gen-attrdef-decls)
mlir_tablegen(SYCLAttributes.cpp.inc -gen-attrdef-defs)
mlir_tablegen(SYCLEnums.h.inc -gen-enum-decls)
mlir_tablegen(SYCLEnums.cpp.inc -gen-enum-defs)
```

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 6 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2333]: I don't think that makes any difference here, as this only affects the width of the number encoding the memory space. We can even use 32 bits. WDYT?

Comment (whitneywhtsang) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2333]: True, approved!

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**